### PR TITLE
Trigger bitbucket rest api event

### PR DESCRIPTION
**Pull request by bush4u**

### CODE REVIEW

**Comment by dylburger** [comment]
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:

- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible
- [ ] [Create subscription webhooks sources (vs polling sources) whenever possible](https://pipedream.com/docs/components/guidelines/#webhook-vs-polling-sources)
- [ ] [Always emit a `summary` with events](https://pipedream.com/docs/components/guidelines/#emit-a-summary)
- [ ] [Use a built-in deduping strategy wherever possible](https://pipedream.com/docs/components/guidelines/#deduping)
- [ ] When building webhook sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#webhook-sources)
- [ ] When building polling sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#polling-sources)

**Comment by bush4u** [comment]
The following historical data were skipped due to unclear data source. 

new-pipeline-event
new-repository-event
new-workspace-event
watch-user

**Review by feyzullah — CHANGES_REQUESTED** [review]
Hi @bush4u, it looks great overall. I added a few comments. Can you please check?

File: components/bitbucket/bitbucket.app.mjs [hunk 1]
```diff
@@ -366,10 +370,92 @@ export default {
...
+        branches = [
+          ...branches,
+          ...result,
+        ];
+        if (result.length < 50) {
+          break;
+        }
+      } while (true);
+      return branches;
+    },
```
[the remark this answers is not in the document]
Comment (bush4u) [reply components/bitbucket/bitbucket.app.mjs:395]: The function is actually not in us. I have taken it out

File: components/bitbucket/sources/new-commit-comment/new-commit-comment.mjs [hunk 2]
```diff
@@ -30,6 +30,53 @@ export default {
...
       ];
     },
+    async loadHistoricalData() {
+      const commits = await this.bitbucket.getCommits({
+        workspaceId: this.workspaceId,
+        repositoryId: this.repositoryId,
+        params: {
+          include: this.branchName,
+          page: 1,
+          pagelen: 25,
```
[the remark this answers is not in the document]
Comment (bush4u) [reply components/bitbucket/sources/new-commit-comment/new-commit-comment.mjs]: Done

File: components/bitbucket/sources/new-commit-comment/new-commit-comment.mjs [hunk 3]
```diff
@@ -30,6 +30,53 @@ export default {
...
+            ...response,
+          ];
+          counter++;
+          if (comments.length > 25) {
+            break;
+          }
+          if (counter >= commits.length - 1) {
+            break;
+          }
+        } while (true);
```
[the remark this answers is not in the document]
Comment (bush4u) [reply components/bitbucket/sources/new-commit-comment/new-commit-comment.mjs]: Thanks for the suggestion. 

The issue here is that we need 25 historical data maximum. 

There might be a challenge with this suggestion. 
Imagine there are 25 commits with 30 comments each, the total commit comments would be 750 commit-comments which is not what we want. What we want is 25 historical data of commit-comments.

**Review by feyzullah — DISMISSED** [review]
Thanks, LGTM!

**Comment by DilanAthukorala** [comment]
@bush4u following issues are observed;

1.  "New Branch" source events are only triggered at the point of source creation/deployment for already existing branches. Events received for new branches after the deployment are not reflected.
image




2.  "New Branch" source gets events from other actions such as commits. It should trigger events only for New Branch events. 
image



3.  "New Pipeline Event" source gets an error upon creation when previous data is not available
image




4. "New Repository Events" source gets error upon creation when previous data not available 
image

**Comment by bush4u** [comment]
> @bush4u following issues are observed;
> 
> 1. "New Branch" source events are only triggered at the point of source creation/deployment for already existing branches. Events received for new branches after the deployment are not reflected.
>    image
> 2. "New Branch" source gets events from other actions such as commits. It should trigger events only for New Branch events.
>    image
> 3. "New Pipeline Event" source gets an error upon creation when previous data is not available
>    image
> 4. "New Repository Events" source gets error upon creation when previous data not available
>    image

Hi @DilanAthukorala ,
Thanks for the review. Kindly find my comments below;

Issue 1: I think this is a bug on that was just noticed. I did not work on any of the post deployment events. 
The scope of the the issue I worked on is simply on historical data (Deploy event). 
Would it be okay if a new issue is logged for that and I will pick it up?

Issue 2: Same as Issue 1.

Issue 3 and issue 4: I have fixed these

**Comment by DilanAthukorala** [comment]
@bush4u New Pipeline Events are not emitted and throw the following error
image

**Comment by DilanAthukorala** [comment]
@bush4u Noted, we will track the issues separately as a bug.

**Comment by bush4u** [comment]
> @bush4u Noted, we will track the issues separately as a bug.

Great. Thanks

**Review by lcaresia — APPROVED** [review]
LGTM!

**Pull request merged**